src/profile/main.mo

```diff
@@ -6,195 +6,26 @@
  */
 
-import Array "mo:stdlib/array.mo";
-import Hash "mo:stdlib/hash.mo";
-import Iter "mo:stdlib/iter.mo";
-import Result "mo:stdlib/result.mo";
-import Trie "mo:stdlib/trie.mo";
-import Util "../common/util.mo";
+import Directory "./directory.mo";
+import Types "./types.mo";
 
-actor Profile {
-
-  type Result<Ok, Error> = Result.Result<Ok, Error>;
-  type Trie<Key, Value> = Trie.Trie<Key, Value>;
-
-  /**
-   * The type of a user identifier.
-   */
-  public type UserId = {
-    unbox : [Word8]
-  };
-
-  /**
-   * Test user identifiers for equality.
-   */
-  func eq(x : UserId, y : UserId) : Bool {
-    Array.equals<Word8>(x.unbox, y.unbox, func (xi, yi) {
-      xi == yi
-    })
-  };
-
-  /**
-   * Create a trie key from a user identifier.
-   */
-  func key(x : UserId) : Trie.Key<UserId> {
-    func convert(bytes : [Word8]) : [Word32] {
-      Array.map<Word8, Word32>(Util.convWord8ToWord32, bytes)
-    };
-    {
-      key = x;
-      hash = Hash.BitVec.hashWord8s(convert(x.unbox))
-    }
-  };
-
-  /**
-   * The type of a profile.
-   */
-  public type Profile = {
-    firstName : [Word8];
-    lastName : [Word8];
-    title : [Word8];
-    company : [Word8];
-    experience : [Word8]
-  };
-
-  /**
-   * Decode a profile.
-   */
-  func decodeProfile(next : () -> ?Word8) : Result<Profile, Text> {
-    Util.decodeWord8ArrayOfWord8SizeThen<Profile>(next, func (bytes0) {
-      Util.decodeWord8ArrayOfWord8SizeThen<Profile>(next, func (bytes1) {
-        Util.decodeWord8ArrayOfWord8SizeThen<Profile>(next, func (bytes2) {
-          Util.decodeWord8ArrayOfWord8SizeThen<Profile>(next, func (bytes3) {
-            Util.decodeWord8ArrayOfWord16SizeThen<Profile>(next, func (bytes4) {
-              #ok({
-                firstName = bytes0;
-                lastName = bytes1;
-                title = bytes2;
-                company = bytes3;
-                experience = bytes4
-              })
-            })
-          })
-        })
-      })
-    })
-  };
-
-  /**
-   * The type of the profile database.
-   */
-  type ProfileDatabase = Trie<UserId, Profile>;
+type Profile = Types.Profile;
+type PrincipalId = Types.PrincipalId;
 
-  /**
-   * Initialize the profile database.
-   */
-  var profiles : ProfileDatabase = Trie.empty<UserId, Profile>();
-
-  /**
-   * Find a profile in the profile database.
-   */
-  public func find(userId : UserId) : async (?Profile) {
-    Trie.find<UserId, Profile>(profiles, key(userId), eq)
-  };
+actor Profile {
+  var directory : Directory.Directory = Directory.Directory();
 
-  /**
-   * Insert a profile into the profile database.
-   */
-  func insert(userId : UserId, profile : Profile) {
-    profiles := Trie.insert<UserId, Profile>(
-      profiles,
-      key(userId),
-      eq,
-      profile
-    ).0
+  public shared { caller } func set (profile : Profile) : async PrincipalId {
+    let userId : PrincipalId = hashBlob(caller);
+    directory.set(userId, profile);
+    userId
```

Comment:
Need to add support for admin roles here so we can bulk add profiles for the demo. Alternatively, we could add that in the canister itself.